### PR TITLE
Return tasks from network plugin methods instead of accepting callbacks

### PR DESCRIPTION
**Pull request by Giorgi**
This is a PR for #85 . This is a breaking change as I had to change the interface of the network plugin.

### CODE REVIEW

File: Network/MvvmCross.Plugins.Network.Test/SimpleRestTest.cs [hunk 1]
```diff
@@ -39,9 +40,7 @@ public void GetDataFromGoogleBooks()
             var request = new MvxRestRequest(url);
             MvxDecodedRestResponse<BookSearchResult> theResponse = null;
             Exception exception = null;
-            client.MakeRequestFor<BookSearchResult>(request,
-                                                    (result) => { theResponse = result; },
-                                                    (error) => { exception = error; });
+            theResponse = await client.MakeRequestFor<BookSearchResult>(request);
 
             System.Threading.Thread.Sleep(3000);
```
Comment (alexshikov) [inline Network/MvvmCross.Plugins.Network.Test/SimpleRestTest.cs]: There is no need for Sleep() anymore

File: Network/MvvmCross.Plugins.Network/Rest/MvxJsonRestClient.cs [hunk 2]
```diff
@@ -18,24 +19,23 @@ public class MvxJsonRestClient
     {
         public Func<IMvxJsonConverter> JsonConverterProvider { get; set; }
 
-        public IMvxAbortable MakeRequestFor<T>(MvxRestRequest restRequest, Action<MvxDecodedRestResponse<T>> successAction, Action<Exception> errorAction)
+        public async Task<MvxDecodedRestResponse<T>> MakeRequestFor<T>(MvxRestRequest restRequest)
         {
-            return MakeRequest(restRequest, (MvxStreamRestResponse streamResponse) =>
+            var streamResponse = await MakeStreamRequest(restRequest);
```
Comment (alexshikov) [inline Network/MvvmCross.Plugins.Network/Rest/MvxJsonRestClient.cs]: Use .ConfigureAwait(false);

**Comment by Giorgi** [comment]
I think we also don't need to use `TryCatch` method to make calls because it's ok if the exception propagates back to the caller.

**Comment by alexshikov** [comment]
I thought about it too... Make sense to refactor this part also.

**Comment by Cheesebaron** [comment]
Things that I don't like and need to be fixed:
- [ ] Methods that are Async need Async postfix. 
- [ ] A lot of methods are kind of lying about being async and are using TaskCompletionSource.

**Comment by Giorgi** [comment]
@Cheesebaron I agree about Async postfix but I have a question about TaskCompletionSource: The implementations for the async methods use `BeginGetResponse` `EndGetResponse` so it shouldn't block the caller. Do you mean that `BeginGetResponse` and `EndGetResponse`  use a threadpool thread to run the request and aren't truly async?

**Comment by alexshikov** [comment]
Does it make sense to migrate to HttpClient? It would open opportunity to use ModerHttpClient.

**Comment by Cheesebaron** [comment]
A move to HttpClient would be great. ModernHttpClient is not really needed anymore with the native Handlers Xamarin are pushing out. So soon it will be obsolete. But sure, if we could provide a way to inject our own handler into it, that would be great.

**Comment by Giorgi** [comment]
Yes, moving to HttpClient would be the best but that means completely rewriting the plugin.

**Comment by Giorgi** [comment]
Async postfix added to the methods.

**Comment by Giorgi** [comment]
Is there anything else I can do as part of this PR?

File: Network/MvvmCross.Plugins.Network/Rest/MvxJsonRestClient.cs [hunk 3]
```diff
@@ -20,9 +20,9 @@ public class MvxJsonRestClient
     {
         public Func<IMvxJsonConverter> JsonConverterProvider { get; set; }
 
-        public async Task<MvxDecodedRestResponse<T>> MakeRequestFor<T>(MvxRestRequest restRequest, CancellationToken cancellationToken = default(CancellationToken))
+        public async Task<MvxDecodedRestResponse<T>> MakeRequestForAsync<T>(MvxRestRequest restRequest, CancellationToken cancellationToken = default(CancellationToken))
         {
-            var streamResponse = await MakeStreamRequest(restRequest, cancellationToken);
+            var streamResponse = await MakeStreamRequestAsync(restRequest, cancellationToken);
```
Comment (alexshikov) [inline Network/MvvmCross.Plugins.Network/Rest/MvxJsonRestClient.cs]: Use `.ConfigureAwait(false)`

Comment (Giorgi) [inline Network/MvvmCross.Plugins.Network/Rest/MvxJsonRestClient.cs]: Done

**Comment by alexshikov** [comment]
Migration to HttpClient can be made with another PR. But there is still concern from @Cheesebaron :

> A lot of methods are kind of lying about being async and are using TaskCompletionSource.

**Comment by slodge** [comment]
just wondering... why did this have to be a breaking change? 

could the old interface and non async code be left as-is and new interfaces and methods be added? this could prevent existing code from breaking?

**Comment by Giorgi** [comment]
@slodge You're right, we can mark current methods as `Obsolete` and add the new ones.

@SeeD-Seifer I can't do anything about it because the underlying webrequest class only has event based async methods.

**Comment by Giorgi** [comment]
@slodge I added the old overloads so there it's not a breaking change any more.

**Comment by martijn00** [comment]
Since this isn't breaking anymore, i think it is save to merge. Thanks for contributing!

**Comment by slodge** [comment]
:+1:  Good job - thanks @Giorgi

**Pull request merged**